### PR TITLE
Add labels to Gltf Node and Mesh assets

### PR DESCRIPTION
**Pull request by freiksenet**
# Objective

Add labels to GltfNode and GltfMesh - they are missing from the assets even though they are need if one wants to write a custom Gltf spawning logic.

Eg AnimationPlayer relies on Name component of the node entities to control the animation. There is no way to actually get names of the gltf nodes, thus you can't manually spawn subtree from the scene and animate it.

## Solution

- Add label field and make use of existing label creation logic to store it there.

## Testing

- Ran all tests
- Fixed tests for node_hierarchy to use lable now

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
This is a nice simple change, and the code quality benefits are a great side effect.

**Comment by mockersf** [comment]
doesn't do what it says, it always uses the auto generated label

**Comment by freiksenet** [comment]
Ok, fixed the logic now. I guess the following can be a possible follow-up:

1. Add label to other "GLTF" named types.
2. Consider wrapping stuff like StandardMaterial or Scene in a Gltf wrapper that has same label/name logic.
3. Add backlinks to parents to all nodes and possibly backlink to scene - this can be problematic because the code operates partially on gltf::Node and partially on GltfNode types (confusing).

**Comment by freiksenet** [comment]
I think I'll try to prototype a different kind of bevy gltf plugin for my usages, that doesn't do this work in loaders and let's see how it goes. I think this PR is still good tho.

File: crates/bevy_gltf/src/lib.rs [hunk 1]
```diff
@@ -121,18 +125,86 @@ pub struct GltfNode {
...
+            name: node.name().map(|s| s.to_string()),
+            children,
+            mesh,
+            transform,
+            extras,
+        }
+    }
+
+    /// Computed name for a node - either a user defined name or a generated name from index
+    pub fn name(&self) -> String {
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:148]: I would prefer those to not return owned string, but not sur that's easy

File: crates/bevy_gltf/src/lib.rs [hunk 2]
```diff
@@ -111,6 +111,10 @@ pub struct Gltf {
 /// See [the relevant glTF specification section](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#reference-node).
 #[derive(Asset, Debug, Clone, TypePath)]
 pub struct GltfNode {
+    /// A user defined node name from GLTF
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:114]:
```suggestion
    /// A user defined node name from glTF
```

File: crates/bevy_gltf/src/lib.rs [hunk 3]
```diff
@@ -121,18 +125,86 @@ pub struct GltfNode {
     pub extras: Option<GltfExtras>,
 }
 
+impl GltfNode {
+    /// Create a node extracting name and index from gltf def
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:129]:
```suggestion
    /// Create a node extracting name and index from glTF def
```

File: crates/bevy_gltf/src/lib.rs [hunk 4]
```diff
@@ -121,18 +125,86 @@ pub struct GltfNode {
...
+    }
+}
+
 /// A glTF mesh, which may consist of multiple [`GltfPrimitives`](GltfPrimitive)
 /// and an optional [`GltfExtras`].
 ///
 /// See [the relevant glTF specification section](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#reference-mesh).
 #[derive(Asset, Debug, Clone, TypePath)]
 pub struct GltfMesh {
+    /// A user defined mesh name from GLTF
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:168]:
```suggestion
    /// A user defined mesh name from glTF
```

File: crates/bevy_gltf/src/lib.rs [hunk 5]
```diff
@@ -121,18 +125,86 @@ pub struct GltfNode {
...
+    /// Index of the mesh inside the scene
+    pub index: usize,
     /// Primitives of the glTF mesh.
     pub primitives: Vec<GltfPrimitive>,
     /// Additional data.
     pub extras: Option<GltfExtras>,
 }
 
+impl GltfMesh {
+    /// Create a mesh extracting name and index from gltf def
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:179]:
```suggestion
    /// Create a mesh extracting name and index from glTF def
```

File: crates/bevy_gltf/src/lib.rs [hunk 6]
```diff
@@ -121,18 +125,86 @@ pub struct GltfNode {
...
+        if let Some(name) = &self.name {
+            name.clone()
+        } else {
+            format!("GltfNode{}", self.index)
+        }
+    }
+
+    /// Computed asset label for node based on its index in the scene
+    pub fn label(&self) -> String {
+        format!("Node{}", self.index)
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:158]: could you comment on your choice to not go for the same string in both those cases?

**Review by Olle-Lukowski — APPROVED** [review]
Other than the concerns that were already mentioned, this looks good to me!

File: crates/bevy_gltf/src/lib.rs (hunk 6 above)
Comment (freiksenet) [reply crates/bevy_gltf/src/lib.rs:158]: This is the asset label within GLTF asset - I will rename for clarity.

File: crates/bevy_gltf/src/lib.rs (hunk 1 above)
Comment (freiksenet) [reply crates/bevy_gltf/src/lib.rs:148]: I can make name/asset_label be precomputed on construction - but that mean it won't update if the object is mutated. That's probably okay tho.

Comment (freiksenet) [reply crates/bevy_gltf/src/lib.rs:148]: Added a commit with that option, can revert it.

**Comment by alice-i-cecile** [comment]
Can you resolve merge conflicts?

**Comment by freiksenet** [comment]
@mockersf Addressed the issues, ported to new GltfAssetLabel system, added name/index/label for primitive. Ready for review.

File: crates/bevy_gltf/src/lib.rs [hunk 7]
```diff
@@ -205,6 +205,16 @@ pub struct Gltf {
...
+    /// Index of the node inside the scene
+    pub index: usize,
+    /// Computed name for a node - either a user defined node name from gLTF or a generated name from index
+    pub name: String,
+    /// Subasset label for this node within the gLTF parent asset.
+    ///
+    /// ```
+    /// asset_server.load(format!("my_glft.glft#{}", node.asset_label());
+    /// ```
+    pub asset_label: String,
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:217]:
```suggestion
    pub asset_label: GltfAssetLabel,
```

File: crates/bevy_gltf/src/lib.rs [hunk 8]
```diff
@@ -215,18 +225,74 @@ pub struct GltfNode {
...
+    pub fn new(
+        node: &gltf::Node,
+        children: Vec<GltfNode>,
+        mesh: Option<Handle<GltfMesh>>,
+        transform: bevy_transform::prelude::Transform,
+        extras: Option<GltfExtras>,
+    ) -> Self {
+        Self {
+            index: node.index(),
+            asset_label: format!("Node{}", node.index()),
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:239]:
```suggestion
            asset_label: GltfAssetLabel::Node(node.index()),
```

File: crates/bevy_gltf/src/lib.rs [hunk 9]
```diff
@@ -215,18 +225,74 @@ pub struct GltfNode {
...
+    /// Index of the mesh inside the scene
+    pub index: usize,
+    /// Computed name for a mesh - either a user defined mesh name from gLTF or a generated name from index
+    pub name: String,
+    /// Subasset label for this mesh within the gLTF parent asset.
+    ///
+    /// ```
+    /// asset_server.load(format!("my_glft.glft#{}", mesh.asset_label());
+    /// ```
+    pub asset_label: String,
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:268]:
```suggestion
    pub asset_label: GltfAssetLabel,
```

File: crates/bevy_gltf/src/lib.rs [hunk 10]
```diff
@@ -215,18 +225,74 @@ pub struct GltfNode {
...
+impl GltfMesh {
+    /// Create a mesh extracting name and index from glTF def
+    pub fn new(
+        mesh: &gltf::Mesh,
+        primitives: Vec<GltfPrimitive>,
+        extras: Option<GltfExtras>,
+    ) -> Self {
+        Self {
+            index: mesh.index(),
+            asset_label: format!("Mesh{}", mesh.index()),
```
Comment (mockersf) [inline crates/bevy_gltf/src/lib.rs:284]:
```suggestion
            asset_label: GltfAssetLabel::Mesh(mesh.index()),
```

File: crates/bevy_gltf/src/loader.rs [hunk 11]
```diff
@@ -1996,8 +1987,11 @@ mod test {
     use crate::GltfNode;
 
     impl GltfNode {
-        fn empty() -> Self {
+        fn with_generated_name(index: usize) -> Self {
             GltfNode {
+                index,
+                asset_label: format!("Node{}", index),
```
Comment (mockersf) [inline crates/bevy_gltf/src/loader.rs:1993]:
```suggestion
                asset_label: GltfAssetLabel::Node(index),
```

**Review by mockersf — CHANGES_REQUESTED** [review]

**Comment by freiksenet** [comment]
@mockersf Apologies, one change didn't get added.

**Comment by freiksenet** [comment]
@mockersf Added an issue with follow-up work, would appreciate your input. https://github.com/bevyengine/bevy/issues/13681

**Review by mockersf — APPROVED** [review]
I would prefer if the `asset_label` fields were functions instead, as is they could have incoherent data with the index fields

**Comment by alice-i-cecile** [comment]
I'm going to merge this for now, but I think that Francois's comment above has merit.

**Pull request merged**